File: src/lib/pipeline/parsers/rc-parser.ts
/**
 * rc-parser.ts — RingCentral phone data parser.
 *
 * Port of Python's rc_parser.py.
 *
 * Handles two file formats:
 *   1. Daily scheduled report (from Outlook): has "Users" sheet with data,
 *      "Filters" sheet with date. No "Date" column in the data.
 *   2. Multi-date export (rc_export.xlsx): has a "Date" column per row.
 *
 * Output rows: { Date, Agent, Calls, Inbound, Outbound, TalkTimeSeconds }
 */

import * as XLSX from "xlsx";
import type { ParseResult } from "../types";
import { Spine } from "../spine";

// ─── Column mapping (original name → internal name) ───
const COL_MAP: Record<string, string> = {
  "Total Calls": "Calls",
  "# Inbound": "Inbound",
  "# Outbound": "Outbound",
  "Total Handle Time": "TalkTime",
};

/**
 * Parse an RC file buffer (Excel workbook).
 *
 * @param fileBuffer  Raw file bytes
 * @param fileName    Original filename (used for date extraction fallback)
 * @param spine       Agent name resolver
 * @param targetDate  ISO date string ("2026-03-22") or null
 */
export function parseRC(
  fileBuffer: ArrayBuffer,
  fileName: string,
  spine: Spine,
  targetDate: string | null,
): ParseResult {
  const logs: string[] = [];
  logs.push(`[rc-parser] Parsing file: ${fileName}`);

  const wb = XLSX.read(fileBuffer, { type: "array", cellDates: true });
  const sheetNames = wb.SheetNames;

  if (sheetNames.includes("Users")) {
    logs.push("[rc-parser] Detected daily report format (Users sheet present)");
    return parseDailyReport(wb, fileName, spine, targetDate, logs);
  } else {
    logs.push("[rc-parser] Detected multi-date export format");
    return parseMultiDate(wb, fileName, spine, targetDate, logs);
  }
}

// ─── Daily report format ────────────────────────────────────────────

function parseDailyReport(
  wb: XLSX.WorkBook,
  fileName: string,
  spine: Spine,
  targetDate: string | null,
  logs: string[],
): ParseResult {
  // 1. Extract date from the Filters sheet's "From Time" column
  let reportDate: string | null = targetDate;

  try {
    const filtersSheet = wb.Sheets["Filters"];
    if (filtersSheet) {
      const filtersData = XLSX.utils.sheet_to_json<Record<string, unknown>>(filtersSheet);
      if (filtersData.length > 0) {
        const firstRow = filtersData[0];
        const fromTime = firstRow["From Time"];
        if (fromTime != null) {
          reportDate = toDateString(fromTime);
          logs.push(`[rc-parser] Extracted date from Filters sheet: ${reportDate}`);
        }
      }
    }
  } catch {
    logs.push("[rc-parser] Could not read Filters sheet for date");
  }

  // 2. Fallback: extract date from filename (e.g. Office_Perf_Users_03_23_2026_...)
  if (!reportDate) {
    const match = fileName.match(/(\d{2})_(\d{2})_(\d{4})/);
    if (match) {
      const [, m, d, y] = match;
      reportDate = `${y}-${m}-${d}`;
      logs.push(`[rc-parser] Extracted date from filename: ${reportDate}`);
    }
  }

  // 3. Last resort: today's date
  if (!reportDate) {
    reportDate = new Date().toISOString().slice(0, 10);
    logs.push(`[rc-parser] Using today's date as fallback: ${reportDate}`);
  }

  // Read the Users sheet
  const usersSheet = wb.Sheets["Users"];
  if (!usersSheet) {
    logs.push("[rc-parser] ERROR: Users sheet not found");
    return { type: "rc", rows: [], logs };
  }

  const data = XLSX.utils.sheet_to_json<Record<string, unknown>>(usersSheet);
  return processRows(data, spine, reportDate, fileName, logs);
}

// ─── Multi-date export format ───────────────────────────────────────

function parseMultiDate(
  wb: XLSX.WorkBook,
  fileName: string,
  spine: Spine,
  targetDate: string | null,
  logs: string[],
): ParseResult {
  const sheet = wb.Sheets[wb.SheetNames[0]];
  let data = XLSX.utils.sheet_to_json<Record<string, unknown>>(sheet);

  // Filter to targetDate if a Date column exists and targetDate is provided
  if (targetDate) {
    const hasDateCol = data.length > 0 && "Date" in data[0];
    if (hasDateCol) {
      data = data.filter((row) => {
        const rowDate = toDateString(row["Date"]);
        return rowDate === targetDate;
      });
      logs.push(`[rc-parser] Filtered to target date ${targetDate}: ${data.length} rows`);
    }
  }

  // For multi-date, we pass null so processRows preserves per-row dates
  const dateForAllRows = targetDate;
  return processRows(data, spine, dateForAllRows, fileName, logs);
}

// ─── Shared processing ─────────────────────────────────────────────

function processRows(
  data: Record<string, unknown>[],
  spine: Spine,
  reportDate: string | null,
  sourceName: string,
  logs: string[],
): ParseResult {
  const rows: Record<string, unknown>[] = [];

  for (const raw of data) {
    // Rename columns via COL_MAP
    const row: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(raw)) {
      const mapped = COL_MAP[key] ?? key;
      row[mapped] = value;
    }

    // Resolve agent name
    const nameVal = row["Name"];
    if (nameVal == null || String(nameVal).trim() === "") continue;

    const agent = spine.resolveAgent(String(nameVal));
    if (!agent) {
      logs.push(`[rc-parser] Skipping unresolved name: "${nameVal}"`);
      continue;
    }

    // Date: use reportDate or fall back to per-row Date column
    let date = reportDate;
    if (!date && row["Date"] != null) {
      date = toDateString(row["Date"]);
    }

    // Numeric columns: default to 0
    const calls = toInt(row["Calls"]);
    const inbound = toInt(row["Inbound"]);
    const outbound = toInt(row["Outbound"]);
    const talkTimeSeconds = toSeconds(row["TalkTime"]);

    rows.push({
      Date: date,
      Agent: agent,
      Calls: calls,
      Inbound: inbound,
      Outbound: outbound,
      TalkTimeSeconds: talkTimeSeconds,
    });
  }

  logs.push(`[rc-parser] Parsed ${rows.length} rows from ${sourceName}`);
  return { type: "rc", rows, logs };
}

// ─── Helper: convert any value to a date string ────────────────────

function toDateString(val: unknown): string | null {
  if (val == null) return null;

  // Already a Date object (SheetJS cellDates:true)
  if (val instanceof Date) {
    return val.toISOString().slice(0, 10);
  }

  // String date — try ISO parse
  const s = String(val).trim();
  if (!s) return null;

  // ISO format "2026-03-22" or "2026-03-22T..."
  const isoMatch = s.match(/^(\d{4}-\d{2}-\d{2})/);
  if (isoMatch) return isoMatch[1];

  // US format "03/22/2026"
  const usMatch = s.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})/);
  if (usMatch) {
    const [, mm, dd, yyyy] = usMatch;
    return `${yyyy}-${mm.padStart(2, "0")}-${dd.padStart(2, "0")}`;
  }

  // Fallback: let JS Date parse it
  const d = new Date(s);
  if (!isNaN(d.getTime())) return d.toISOString().slice(0, 10);

  return null;
}

// ─── Helper: convert TalkTime to total seconds ─────────────────────

/**
 * Convert a TalkTime value to total seconds.
 * Handles:
 *   - Excel serial time (fractional day number, e.g. 0.00381944 ≈ 330s)
 *   - Timedelta strings from pandas ("0 days 00:05:30")
 *   - Plain HH:MM:SS or MM:SS strings ("00:05:30")
 *   - Already-numeric seconds
 *   - Date objects (treated as time-of-day offset from midnight)
 */
function toSeconds(val: unknown): number {
  if (val == null) return 0;

  // Excel serial time: a number between 0 and 1 (fraction of a day)
  // Numbers >= 1 are also valid (e.g. 1.5 = 36 hours)
  if (typeof val === "number") {
    if (val === 0) return 0;
    // If it looks like an Excel serial time (fractional day),
    // convert: 1 day = 86400 seconds
    if (val > 0 && val < 500) {
      // Treat values < 500 as fractional days (up to ~500 days).
      // Exact integers might be raw seconds, but Excel time values
      // for call durations are almost always < 1.
      if (val < 1) {
        return Math.round(val * 86400);
      }
      // If it's a small integer, it could be seconds already or a fractional day.
      // Excel serial times for talk time are typically < 1, so integers > 1
      // are likely already in seconds.
      return Math.round(val);
    }
    return Math.round(val);
  }

  // Date object from SheetJS cellDates:true — extract time-of-day
  if (val instanceof Date) {
    const hours = val.getUTCHours();
    const minutes = val.getUTCMinutes();
    const seconds = val.getUTCSeconds();
    return hours * 3600 + minutes * 60 + seconds;
  }

  // String formats
  if (typeof val === "string") {
    const s = val.trim();
    if (!s) return 0;

    // Pandas timedelta string: "0 days 00:05:30" or "0 days 00:05:30.000000"
    const tdMatch = s.match(/(\d+)\s+days?\s+(\d{1,2}):(\d{2}):(\d{2})/);
    if (tdMatch) {
      const [, days, hh, mm, ss] = tdMatch;
      return (
        parseInt(days) * 86400 +
        parseInt(hh) * 3600 +
        parseInt(mm) * 60 +
        parseInt(ss)
      );
    }

    // HH:MM:SS
    const hmsMatch = s.match(/^(\d{1,2}):(\d{2}):(\d{2})$/);
    if (hmsMatch) {
      const [, hh, mm, ss] = hmsMatch;
      return parseInt(hh) * 3600 + parseInt(mm) * 60 + parseInt(ss);
    }

    // MM:SS
    const msMatch = s.match(/^(\d{1,2}):(\d{2})$/);
    if (msMatch) {
      const [, mm, ss] = msMatch;
      return parseInt(mm) * 60 + parseInt(ss);
    }

    // Numeric string
    const num = parseFloat(s);
    if (!isNaN(num)) {
      if (num > 0 && num < 1) return Math.round(num * 86400);
      return Math.round(num);
    }

    return 0;
  }

  return 0;
}

// ─── Helper: safe int parse ────────────────────────────────────────

function toInt(val: unknown): number {
  if (val == null) return 0;
  if (typeof val === "number") return isNaN(val) ? 0 : Math.round(val);
  const n = parseInt(String(val), 10);
  return isNaN(n) ? 0 : n;
}

File: automation/src/parsers/rc_parser.py
"""
rc_parser.py — RingCentral phone data parser.

Handles two file formats:
  1. Multi-date export (has Date column) — used for historical data (rc_export.xlsx)
  2. Daily scheduled report from Outlook (no Date column, 3 sheets: Filters, KPIs, Users)
     - Date is extracted from the Filters sheet's "From Time" column
     - Data is on the "Users" sheet

Output: DataFrame with columns:
  Date, Agent, Calls, Inbound, Outbound, TalkTimeSeconds
"""

import re
import pandas as pd
from pathlib import Path
from datetime import timedelta, date as date_type

from src.spine import Spine


def parse(file_path: str, spine: Spine, target_date=None, sheet_name=None) -> pd.DataFrame:
    """
    Parse an RC export file.

    Parameters
    ----------
    file_path : str
        Path to the RC Excel file.
    spine : Spine
        Agent name resolver.
    target_date : date, optional
        Filter to this date only. For daily reports (no Date column),
        this is used as the date for all rows.
    sheet_name : str or int, optional
        Sheet to read. If None, auto-detects format.
    """
    path = Path(file_path)

    # Auto-detect format: check if "Users" sheet exists
    try:
        import openpyxl
        wb = openpyxl.load_workbook(str(path), read_only=True)
        sheets = wb.sheetnames
        wb.close()
    except Exception:
        sheets = []

    if "Users" in sheets and sheet_name is None:
        # Daily scheduled report format (from Outlook)
        return _parse_daily_report(file_path, spine, target_date)
    else:
        # Multi-date export format (historical)
        return _parse_multi_date(file_path, spine, target_date, sheet_name or 0)


def _parse_daily_report(file_path: str, spine: Spine, target_date=None) -> pd.DataFrame:
    """Parse the daily scheduled report (Filters + Users sheets)."""
    # Extract date from Filters sheet
    report_date = target_date
    try:
        filters_df = pd.read_excel(file_path, sheet_name="Filters", engine="openpyxl")
        if "From Time" in filters_df.columns:
            from_time = filters_df["From Time"].dropna().iloc[0]
            report_date = pd.to_datetime(from_time).date()
    except Exception:
        pass

    if report_date is None:
        # Try to extract date from filename (e.g. Office_Perf_Users_03_23_2026_...)
        import re
        match = re.search(r'(\d{2})_(\d{2})_(\d{4})', Path(file_path).name)
        if match:
            m, d, y = match.groups()
            report_date = date_type(int(y), int(m), int(d))
        else:
            from datetime import date
            report_date = date.today()

    # Read Users sheet
    df = pd.read_excel(file_path, sheet_name="Users", engine="openpyxl")
    return _process_df(df, spine, report_date, Path(file_path).name)


def _parse_multi_date(file_path: str, spine: Spine, target_date=None,
                      sheet_name=0) -> pd.DataFrame:
    """Parse the multi-date export format (has Date column)."""
    df = pd.read_excel(file_path, engine="openpyxl", sheet_name=sheet_name)

    if "Date" in df.columns:
        df["Date"] = pd.to_datetime(df["Date"], errors="coerce")
        if target_date is not None:
            df = df[df["Date"].dt.date == pd.Timestamp(target_date).date()]

    report_date = target_date  # None means _process_df will keep per-row dates
    return _process_df(df, spine, report_date, Path(file_path).name)


def _process_df(df: pd.DataFrame, spine: Spine, report_date, source_name: str = "") -> pd.DataFrame:
    """Common processing for both formats."""
    col_map = {
        "Total Calls": "Calls",
        "# Inbound": "Inbound",
        "# Outbound": "Outbound",
        "Total Handle Time": "TalkTime",
    }
    df = df.rename(columns={k: v for k, v in col_map.items() if k in df.columns})

    # Resolve agent names
    df["Agent"] = df["Name"].apply(spine.resolve_agent)
    df = df.dropna(subset=["Agent"])

    # Convert TalkTime to total seconds
    df["TalkTimeSeconds"] = df["TalkTime"].apply(_to_seconds) if "TalkTime" in df.columns else 0

    # Set date (preserve per-row dates in multi-date mode)
    if report_date is not None:
        df["Date"] = report_date
    elif "Date" not in df.columns:
        df["Date"] = None

    # Build result
    for col in ["Calls", "Inbound", "Outbound"]:
        if col in df.columns:
            df[col] = df[col].fillna(0).astype(int)
        else:
            df[col] = 0

    result = df[["Date", "Agent", "Calls", "Inbound", "Outbound", "TalkTimeSeconds"]].copy()
    result = result.fillna(0)

    print(f"[rc_parser] Parsed {len(result)} rows{f' from {source_name}' if source_name else ''}")
    return result


def parse_for_date(
    raw_folder: str,
    spine: Spine,
    target_date: date_type,
) -> pd.DataFrame | None:
    """
    Find and parse the RC file whose DATA date matches target_date.

    RC filename convention: rc_YYYYMMDD_Office_Perf_Users_...
    The date in the filename is the PULL date (email arrival date).
    The actual DATA date = pull_date - 1 day.

    So rc_20260429 contains data for 2026-04-28.

    Returns DataFrame or None if no matching file exists.
    """
    folder = Path(raw_folder)
    if not folder.exists():
        return None

    rc_files = list(folder.glob("rc_*.xlsx")) + list(folder.glob("rc_*.xls"))
    if not rc_files:
        print(f"[rc_parser] No RC files found in {raw_folder}")
        return None

    # Find the file whose data date matches target_date
    for f in rc_files:
        data_date = _data_date_from_filename(f)
        if data_date == target_date:
            print(f"[rc_parser] Found RC file for {target_date}: {f.name}")
            return parse(str(f), spine, target_date=target_date)

    # No exact match — report clearly (don't fall back to wrong file)
    newest = max(rc_files, key=lambda f: f.stat().st_mtime)
    newest_data_date = _data_date_from_filename(newest)
    print(f"[rc_parser] No RC file for {target_date} "
          f"(newest has data for {newest_data_date}: {newest.name})")
    return None


def _data_date_from_filename(path: Path) -> date_type | None:
    """
    Extract the DATA date from an RC filename.

    Filename: rc_YYYYMMDD_Office_Perf_Users_...
    The YYYYMMDD is the PULL date. Data date = pull_date - 1 day.
    """
    m = re.search(r"rc_(\d{4})(\d{2})(\d{2})", path.name)
    if m:
        try:
            pull_date = date_type(int(m.group(1)), int(m.group(2)), int(m.group(3)))
            return pull_date - timedelta(days=1)
        except ValueError:
            return None
    return None


def _to_seconds(val) -> int:
    """Convert timedelta or string duration to total seconds."""
    if isinstance(val, timedelta):
        return int(val.total_seconds())
    if pd.isna(val):
        return 0
    # Handle string format like "0 days 00:05:30"
    if isinstance(val, str):
        try:
            return int(pd.to_timedelta(val).total_seconds())
        except Exception:
            return 0
    return 0
